### PR TITLE
Use Pathlib in test_application.py and test_completerlib.py

### LINKED ISSUE #12515: Pathlib, Pathlib everywhere.
There are many places where we could make use of Pathlib. 

Look for any places that uses `with open(...)` and ask yourself:

- is the argument a string ? 
- would it make sens to make it a `Path()`, 
- how far upstream in the code can I make it a Path.

Don't try to bite more than you can chew (or more than I can review), try to fix 1 place at a time.

### PR DESCRIPTION
**Pull request by kir0ul**
Refactored `test_application.py` to use Pathlib.

Fix #12515.

### CODE REVIEW

File: IPython/core/tests/test_completerlib.py [hunk 1]
```diff
@@ -77,21 +77,22 @@ def test_completion_more_args(self):
...
         print(repr(event.line))
         match = set(magic_run_completer(None, event))
         # We specifically use replace here rather than normpath, because
         # at one point there were duplicates 'adir' and 'adir/', and normpath
         # would hide the failure for that.
-        self.assertEqual(match, {join(self.BASETESTDIR, f).replace('\\','/')
-                            for f in (u'a.py', u'aao.py', u'aao.txt', u'adir/')})
+        self.assertEqual(match, {str(self.BASETESTDIR.joinpath(f)) if Path(f).suffix
+                                 else str(self.BASETESTDIR.joinpath(f)) + os.sep
+                            for f in (u"a.py", u"aao.py", u"aao.txt", u"adir/")})
```
Comment (kir0ul) [inline IPython/core/tests/test_completerlib.py:88]: Not sure it's the best way to do this, it works but seems more complicated than before... :thinking:

Comment (kir0ul) [reply IPython/core/tests/test_completerlib.py]: Indeed it doesn't work on Windows... :sweat:

File: IPython/core/tests/test_application.py [hunk 2]
```diff
@@ -33,15 +33,15 @@ def test_unicode_cwd():
 @dec.onlyif_unicode_paths
 def test_unicode_ipdir():
     """Check that IPython starts with non-ascii characters in the IP dir."""
-    ipdir = tempfile.mkdtemp(suffix=u"€")
+    ipdir = Path(tempfile.mkdtemp(suffix=u"€"))
     
     # Create the config file, so it tries to load it.
-    with open(os.path.join(ipdir, 'ipython_config.py'), "w") as f:
+    with ipdir.joinpath("ipython_config.py").open("w") as f:
         pass
```
Comment (Carreau) [inline IPython/core/tests/test_application.py:40]: might be able to replace with something like https://docs.python.org/3/library/pathlib.html#pathlib.Path.touch

```diff
(ipdir / "ipython_config.py").touch()
```

Comment (kir0ul) [reply IPython/core/tests/test_application.py]: Good catch!

**Comment by Carreau** [comment]
Many thanks for working on this a few years ago; I was never able to keep up with IPython PRs, and this now conflicts, so I'm going to close.

**Pull request closed without merging**